File: vendkit/core/versions.py
"""Version grammar, ordering, channels, retraction awareness.

Stdlib only — ``is-newer`` runs on the consumer PR path (INV-9).
Grammar and semantics: releases-and-versioning spec §2.
"""

from __future__ import annotations

import re

from .util import Refusal, UsageError

_STABLE = re.compile(r"^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$")
_RC = re.compile(r"^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)-rc\.([1-9]\d*)$")

# Ordering key: (major, minor, patch, is_stable, rc_n). A stable release
# sorts above every rc of the same triple (SemVer §11).
Key = tuple[int, int, int, int, int]


def parse(version: str, channel: str = "stable") -> Key | None:
    """Parse a tag name; None if it is not release-shaped for the channel."""
    m = _STABLE.match(version)
    if m:
        a, b, c = (int(g) for g in m.groups())
        return (a, b, c, 1, 0)
    if channel == "rc":
        m = _RC.match(version)
        if m:
            a, b, c, n = (int(g) for g in m.groups())
            return (a, b, c, 0, n)
    return None


def require(version: str, channel: str = "rc") -> Key:
    """Parse accepting the widest grammar; UsageError if malformed."""
    key = parse(version, channel)
    if key is None:
        raise UsageError(f"not a release version: {version!r}")
    return key


def is_newer(pinned: str, target: str, retracted: list[str] | None = None) -> bool:
    """True iff target > pinned. Refuses a retracted target (exit 3)."""
    if retracted and target in retracted:
        raise Refusal("retracted", f"target {target} is retracted by the publisher")
    return require(target) > require(pinned)


def latest(
    tags: list[str], channel: str = "stable", retracted: list[str] | None = None
) -> str | None:
    """Greatest qualifying version among tag names; None when none qualify."""
    best: tuple[Key, str] | None = None
    for name in tags:
        if retracted and name in retracted:
            continue
        key = parse(name, channel)
        if key is None:
            continue
        if best is None or key > best[0]:
            best = (key, name)
    return best[1] if best else None


def classify_bump(pinned: str, target: str) -> str:
    """'patch' | 'minor' | 'major' for the pinned→target jump."""
    p, t = require(pinned), require(target)
    if t[0] != p[0]:
        return "major"
    if t[1] != p[1]:
        return "minor"
    return "patch"


def bump(version: str, kind: str) -> str:
    a, b, c, stable, _ = require(version)
    if not stable:
        raise UsageError(f"cannot bump from a pre-release: {version}")
    if kind == "major":
        return f"v{a + 1}.0.0"
    if kind == "minor":
        return f"v{a}.{b + 1}.0"
    if kind == "patch":
        return f"v{a}.{b}.{c + 1}"
    raise UsageError(f"unknown bump kind: {kind!r}")


def in_window(pinned: str, applies_from: str, target: str) -> bool:
    """Migration window arithmetic: pinned < applies_from <= target."""
    return require(pinned) < require(applies_from) <= require(target)

File: internal/core/versions.go
// Version grammar, ordering, channels, retraction awareness
// (releases-and-versioning spec §2).

package core

import (
	"fmt"
	"regexp"
	"strconv"
)

var (
	stableRx = regexp.MustCompile(`^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$`)
	rcRx     = regexp.MustCompile(`^v(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)-rc\.([1-9]\d*)$`)
)

// VersionKey orders as (major, minor, patch, is_stable, rc_n): a stable
// release sorts above every rc of the same triple (SemVer §11).
type VersionKey struct{ Major, Minor, Patch, Stable, RC int }

func (k VersionKey) Less(o VersionKey) bool {
	a := [5]int{k.Major, k.Minor, k.Patch, k.Stable, k.RC}
	b := [5]int{o.Major, o.Minor, o.Patch, o.Stable, o.RC}
	for i := range a {
		if a[i] != b[i] {
			return a[i] < b[i]
		}
	}
	return false
}

func atoi(s string) int { n, _ := strconv.Atoi(s); return n }

// ParseVersion parses a tag name; ok=false if not release-shaped for the
// channel ("stable" or "rc").
func ParseVersion(version, channel string) (VersionKey, bool) {
	if m := stableRx.FindStringSubmatch(version); m != nil {
		return VersionKey{atoi(m[1]), atoi(m[2]), atoi(m[3]), 1, 0}, true
	}
	if channel == "rc" {
		if m := rcRx.FindStringSubmatch(version); m != nil {
			return VersionKey{atoi(m[1]), atoi(m[2]), atoi(m[3]), 0, atoi(m[4])}, true
		}
	}
	return VersionKey{}, false
}

// RequireVersion parses accepting the widest grammar; *UsageError if malformed.
func RequireVersion(version string) (VersionKey, error) {
	key, ok := ParseVersion(version, "rc")
	if !ok {
		return VersionKey{}, Usagef("not a release version: %q", version)
	}
	return key, nil
}

// IsNewer: true iff target > pinned. Refuses a retracted target (exit 3).
func IsNewer(pinned, target string, retracted []string) (bool, error) {
	for _, r := range retracted {
		if r == target {
			return false, &Refusal{Reason: "retracted",
				Msg: fmt.Sprintf("target %s is retracted by the publisher", target)}
		}
	}
	t, err := RequireVersion(target)
	if err != nil {
		return false, err
	}
	p, err := RequireVersion(pinned)
	if err != nil {
		return false, err
	}
	return p.Less(t), nil
}

// Latest: greatest qualifying version among tag names; "" when none qualify.
func Latest(tags []string, channel string, retracted []string) string {
	best := ""
	var bestKey VersionKey
	for _, name := range tags {
		skip := false
		for _, r := range retracted {
			if r == name {
				skip = true
				break
			}
		}
		if skip {
			continue
		}
		key, ok := ParseVersion(name, channel)
		if !ok {
			continue
		}
		if best == "" || bestKey.Less(key) {
			best, bestKey = name, key
		}
	}
	return best
}

// ClassifyBump: "patch" | "minor" | "major" for the pinned→target jump.
func ClassifyBump(pinned, target string) (string, error) {
	p, err := RequireVersion(pinned)
	if err != nil {
		return "", err
	}
	t, err := RequireVersion(target)
	if err != nil {
		return "", err
	}
	switch {
	case t.Major != p.Major:
		return "major", nil
	case t.Minor != p.Minor:
		return "minor", nil
	default:
		return "patch", nil
	}
}

func BumpVersion(version, kind string) (string, error) {
	k, err := RequireVersion(version)
	if err != nil {
		return "", err
	}
	if k.Stable != 1 {
		return "", Usagef("cannot bump from a pre-release: %s", version)
	}
	switch kind {
	case "major":
		return fmt.Sprintf("v%d.0.0", k.Major+1), nil
	case "minor":
		return fmt.Sprintf("v%d.%d.0", k.Major, k.Minor+1), nil
	case "patch":
		return fmt.Sprintf("v%d.%d.%d", k.Major, k.Minor, k.Patch+1), nil
	}
	return "", Usagef("unknown bump kind: %q", kind)
}

// InWindow: migration window arithmetic — pinned < applies_from <= target.
func InWindow(pinned, appliesFrom, target string) (bool, error) {
	p, err := RequireVersion(pinned)
	if err != nil {
		return false, err
	}
	a, err := RequireVersion(appliesFrom)
	if err != nil {
		return false, err
	}
	t, err := RequireVersion(target)
	if err != nil {
		return false, err
	}
	return p.Less(a) && (a.Less(t) || a == t), nil
}
